### DISCUSSION TITLE
configmap and content aware hashing

### DISCUSSION THREAD

**Discussion by tcurdt** [kubernetes-sigs/kustomize#5711] [Q&A]
I really don't get the config map hashing feature as documented.

What I want:

List all the configmaps in the config that should be renamed to their name with their content hash added as a suffix.

Is this possible or not? The `configMapGenerator` does not really seem to help here.

What I tried

```
resources:
- infra/caddy.yaml

configMapGenerator:
  - name: caddy
    files:
      - infra/caddy.yaml
```

Which already is a bit flawed because I want configmap caddy - but in the infra namespace.
What I would like is to have this modification:

```
apiVersion: v1
kind: ConfigMap
metadata:
  namespace: infra
  name: caddy-HASH
  labels:
    app: caddy
data:
  Caddyfile: |
    ...

...
      volumes:
        - name: caddy-config
          configMap:
            name: caddy-HASH
```

**Comment by ChristianCiach** [comment]
You cannot use Kustomize to add a hash-suffix to an existing ConfigMap that you are providing yourself. Kustomize only adds hash-suffixes to ConfigMaps that are *generated* by Kustomize. 

I assume the file `infra/caddy.yaml` from your example is actually a ConfigMap *containing* the Caddy configuration, right?

What you should do is this:

```yaml
resources:
  - deployment.yaml # or some other workload that consumes the ConfigMap named `caddy-config`

configMapGenerator:
  - name: caddy-config
    files:
      - Caddyfile
```

**Comment by tcurdt** [comment]
> You cannot use Kustomize to add a hash-suffix to an existing ConfigMap that you are providing yourself. Kustomize only adds hash-suffixes to ConfigMaps that are generated by Kustomize.

You can. It is just not documented and uses an internal annotation - but it works.
And it would be nice to turn it into an officially supported feature.

TLDR: This annotation does the trick to apply hashes to existing configmaps.

`internal.config.kubernetes.io/needsHashSuffix: enabled`

Unfortunately the problem is also is a bit more complex than it should be.
Is a merged map a generated one or not?

If I merge a generated map with an existing map the hash does not get applied either.
I really wish this would follow the KISS principle.

https://github.com/kubernetes-sigs/kustomize/issues/5712

> What you should do is this:

That's really not what I want to do though.
